File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/DefenseProficiencyEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum DefenseProficiencyEnum {
    UNARMORED_DEFENSE("unarmored_defense"),
    LIGHT_ARMOR("light_armor"),
    MEDIUM_ARMOR("medium_armor"),
    HEAVY_ARMOR("heavy_armor");

    private final String value;
    private final static Map<String, DefenseProficiencyEnum> CONSTANTS = new HashMap<>();

    static {
        for (DefenseProficiencyEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    DefenseProficiencyEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static DefenseProficiencyEnum fromValue(String value) {
        DefenseProficiencyEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/DefenseProficiencyEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class DefenseProficiencyEnum(private val value: String) {
    UNARMORED_DEFENSE("unarmoredDefense"), LIGHT_ARMOR("lightArmor"), MEDIUM_ARMOR("mediumArmor"), HEAVY_ARMOR("heavyArmor");

    companion object {
        private val CONSTANTS: MutableMap<String, DefenseProficiencyEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): DefenseProficiencyEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}